cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/spec/PipelineSpecGenerator.java

```diff
@@ -303,28 +307,35 @@ private ConfiguredStage configureStage(ETLStage stage, ValidatedPipeline validat
   /**
    * Configures a plugin and returns the spec for it.
    *
-   * @param pluginId the unique plugin id
+   * @param stageName the unique plugin id
    * @param etlPlugin user provided configuration for the plugin
-   * @param pipelineConfigurer default pipeline configurere to configure the plugin
+   * @param pipelineConfigurer default pipeline configurer to configure the plugin
    * @return the spec for the plugin
+   * @throws InvalidPipelineException if the plugin threw an exception during configuration
    */
-  protected PluginSpec configurePlugin(String pluginId, ETLPlugin etlPlugin,
-                                       DefaultPipelineConfigurer<T> pipelineConfigurer) {
+  public StageSpec.Builder configureStage(String stageName, ETLPlugin etlPlugin,
+                                          DefaultPipelineConfigurer<T> pipelineConfigurer)
+    throws InvalidPipelineException {
     TrackedPluginSelector pluginSelector = new TrackedPluginSelector(
-      new ArtifactSelectorProvider(etlPlugin.getType(), etlPlugin.getName())
-        .getPluginSelector(etlPlugin.getArtifactConfig()));
+      new ArtifactSelectorProvider().getPluginSelector(etlPlugin.getArtifactConfig()));
     String type = etlPlugin.getType();
-    Object plugin = configurer.usePlugin(etlPlugin.getType(),
-                                         etlPlugin.getName(),
-                                         pluginId,
-                                         etlPlugin.getPluginProperties(),
-                                         pluginSelector);
 
+    Object plugin;
+    try {
+      plugin = configurer.usePlugin(etlPlugin.getType(),
+                                    etlPlugin.getName(),
+                                    stageName,
+                                    etlPlugin.getPluginProperties(),
+                                    pluginSelector);
+    } catch (Exception e) {
```

Comment:
It can be other types of exceptions. For example, it's possible the class failed to instantiate because of some packaging error in the plugin or something like that. 
Ideally the InvalidPluginConfigException would contain information about which properties are problematic, which would allow us to catch it here and translate it to a more useful error. As it currently is, we can't be any more specific.

Reply:
i see.